File: gremlin-scala/src/test/scala/com/tinkerpop/gremlin/scala/sideeffect/GroupByStepTest.scala
package com.tinkerpop.gremlin.scala.sideeffect

import com.tinkerpop.gremlin.test.ComplianceTest
import com.tinkerpop.blueprints.pgm.impls.tg.TinkerGraphFactory
import com.tinkerpop.blueprints.pgm.Vertex
import com.tinkerpop.gremlin.scala._

/**
 * @author Marko A. Rodriguez (http://markorodriguez.com)
 */

class GroupByStepTest extends com.tinkerpop.gremlin.test.sideeffect.GroupByStepTest {

  val g = TinkerGraphFactory.createTinkerGraph();

  override def testCompliance() {
    ComplianceTest.testCompliance(this.getClass);
  }

  def test_g_V_groupByXlang_nameX() {
    val m = new java.util.HashMap[String, java.util.List[String]]();
    super.test_g_V_groupByXlang_nameX(g.V.groupBy(m.asInstanceOf[java.util.HashMap[_, java.util.List[_]]], {v: Vertex => v("lang")}, {v: Vertex => v("name")}), m);
  }
}

File: gremlin-java/src/test/java/com/tinkerpop/gremlin/java/sideeffect/GroupByStepTest.java
package com.tinkerpop.gremlin.java.sideeffect;

import com.tinkerpop.blueprints.Graph;
import com.tinkerpop.blueprints.Vertex;
import com.tinkerpop.blueprints.impls.tg.TinkerGraphFactory;
import com.tinkerpop.gremlin.java.GremlinPipeline;
import com.tinkerpop.gremlin.test.ComplianceTest;
import com.tinkerpop.pipes.PipeFunction;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Marko A. Rodriguez (http://markorodriguez.com)
 */
public class GroupByStepTest extends com.tinkerpop.gremlin.test.sideeffect.GroupByStepTest {

    Graph g = TinkerGraphFactory.createTinkerGraph();

    public void testCompliance() {
        ComplianceTest.testCompliance(this.getClass());
    }

    public void test_g_V_groupByXlang_nameX() {
        Map<?, List<?>> m = new HashMap<Object, List<?>>();
        super.test_g_V_groupByXlang_nameX(new GremlinPipeline(g.getVertices()).groupBy(m, new PipeFunction<Vertex, String>() {
                    public String compute(Vertex vertex) {
                        return (String) vertex.getProperty("lang");
                    }
                }, new PipeFunction<Vertex, String>() {
                    public String compute(Vertex vertex) {
                        return (String) vertex.getProperty("name");
                    }
                }
        ), (Map) m);

        m = new HashMap<Object, List<?>>();
        super.test_g_V_groupByXlang_nameX(new GremlinPipeline(g.getVertices()).optimize(false).groupBy(m, new PipeFunction<Vertex, String>() {
                    public String compute(Vertex vertex) {
                        return (String) vertex.getProperty("lang");
                    }
                }, new PipeFunction<Vertex, String>() {
                    public String compute(Vertex vertex) {
                        return (String) vertex.getProperty("name");
                    }
                }
        ), (Map) m);
    }

}